File: react-app/src/store/maker.js
import { csrfFetch } from '../helpers';

const LOAD_MAKERS = 'makers/loadMakers';
const CREATE_MAKER = 'makers/createMaker';
const EDIT_MAKER = 'makers/editMaker';
const DELETE_MAKER = 'makers/deleteMaker';

export const loadMakers = () => async (dispatch) => {
    const response = await csrfFetch('/api/makers/');
    const allMakers = await response.json();
    dispatch(loadMakersAction(allMakers));
    return allMakers;
};

const loadMakersAction = (allMakers) => ({
    type: LOAD_MAKERS,
    allMakers
});

export const createMaker = ({
    name,
    website,
    instagram,
    country,
    isYarn,
    isFiber,
    isNotions,
    isPattern,
    isKnitting,
    isCrochet,
    isEmbroidery
}) => async (dispatch) => {
    const response = await csrfFetch('/api/makers/', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            name,
            website,
            instagram,
            country,
            is_yarn: isYarn,
            is_fiber: isFiber,
            is_notions: isNotions,
            is_pattern: isPattern,
            is_knitting: isKnitting,
            is_crochet: isCrochet,
            is_embroidery: isEmbroidery
        })
    });

    if (response.ok) {
        const maker = await response.json();
        dispatch(createMakerAction(maker));
        return maker;
    };
};

const createMakerAction = (maker) => ({
    type: CREATE_MAKER,
    maker
});

export const editMaker = ({
    id,
    name,
    website,
    instagram,
    country,
    isYarn,
    isFiber,
    isNotions,
    isPattern,
    isKnitting,
    isCrochet,
    isEmbroidery
}) => async (dispatch) => {
    const response = await csrfFetch('/api/makers/', {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            id,
            name,
            website,
            instagram,
            country,
            is_yarn: isYarn,
            is_fiber: isFiber,
            is_notions: isNotions,
            is_pattern: isPattern,
            is_knitting: isKnitting,
            is_crochet: isCrochet,
            is_embroidery: isEmbroidery
        })
    });

    if (response.ok) {
        const maker = await response.json();
        dispatch(editMakerAction(maker));
        return maker;
    };
};

const editMakerAction = (maker) => ({
    type: EDIT_MAKER,
    maker
});

export const deleteMaker = (makerId) => async (dispatch) => {
    const response = await csrfFetch('/api/makers/', {
        method: 'DELETE',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ id: makerId })
    });
    const maker = await response.json();
    dispatch(deleteMakerAction(maker, makerId));
};

const deleteMakerAction = (maker, makerId) => {
    return {
        type: DELETE_MAKER,
        maker,
        makerId
    };
};

const makerReducer = (state = {}, action) => {
    let newState = { ...state };

    switch (action.type) {
        case LOAD_MAKERS:
            action.allMakers.forEach(maker => {
                newState[maker.id] = maker;
            });
            return newState;
        case CREATE_MAKER:
        case EDIT_MAKER:
            newState[action.maker.id] = action.maker;
            return newState;
        case DELETE_MAKER:
            delete newState[action.makerId];
            return newState;

        default:
            return state;
    };
};

export default makerReducer;

File: app/models/maker.py
from .db import db



class Maker(db.Model):
    __tablename__ = 'makers'

    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(300), nullable=False, unique=True)
    website = db.Column(db.String(300), nullable=False)
    instagram = db.Column(db.String(100))
    country = db.Column(db.String(4), nullable=False)

    is_yarn = db.Column(db.Boolean, nullable=False)
    is_fiber = db.Column(db.Boolean, nullable=False)
    is_notions = db.Column(db.Boolean, nullable=False)

    is_pattern = db.Column(db.Boolean, nullable=False)
    is_knitting = db.Column(db.Boolean, nullable=False)
    is_crochet = db.Column(db.Boolean, nullable=False)
    is_embroidery = db.Column(db.Boolean, nullable=False)

    def to_dict(self):
        return {
            'id': self.id,
            'name': self.name,
            'website': self.website,
            'instagram': self.instagram,
            'country': self.country,
            'is_yarn': self.is_yarn,
            'is_fiber': self.is_fiber,
            'is_notions': self.is_notions,
            'is_pattern': self.is_pattern,
            'is_knitting': self.is_knitting,
            'is_crochet': self.is_crochet,
            'is_embroidery': self.is_embroidery
        }

    def to_JSON(self):
        return {
            'id': self.id,
            'name': self.name,
            'website': self.website,
            'instagram': self.instagram,
            'country': self.country,
            'isYarn': self.is_yarn,
            'isFiber': self.is_fiber,
            'isNotions': self.is_notions,
            'isPattern': self.is_pattern,
            'isKnitting': self.is_knitting,
            'isCrochet': self.is_crochet,
            'isEmbroidery': self.is_embroidery
        }
